### PR TITLE
improved the doco on the CommandRunner

### PR DESCRIPTION
**Pull request by bsutton**
I've submitted this PR to replace #127 as that PR has languished do to a the lack of a contributor signoff.

### CODE REVIEW

**Comment by bsutton** [comment]
The CI failure here has nothing to do with my changes to the readme. Looks like a bug in the test packages used.

**Comment by bsutton** [comment]
Any chance of getting this pr merged.  
Its non-breaking :)

File: README.md [hunk 1]
```diff
@@ -270,16 +270,39 @@ the left of `"commit"`, it is given to the top-level parser.
...
-as handling `--help` flags and invalid arguments. For example:
+as handling `--help` flags and invalid arguments.
+
+When using the [CommandRunner][] it replaces the [ArgParser][].
+
+In the following example we build a dart application called `dgit` that takes commands `commit` and `stash`.
+
+The [CommandRunner][] takes an 'executableName' which is ONLY used to generate the help message.
+
+> Note: for you final application to be called `dgit` rather than `dgit.dart` you would need to compile it using dart2native.
```
Comment (natebosch) [inline README.md:281]: Drop this line. `dart2native` is not the common approach for this. `pub global activate` is much more common - but in either case it doesn't need to be documented here.

File: README.md [hunk 2]
```diff
@@ -270,16 +270,39 @@ the left of `"commit"`, it is given to the top-level parser.
...
 and [Command][] classes to help structure it. [CommandRunner][] has built-in
 support for dispatching to [Command][]s based on command-line arguments, as well
-as handling `--help` flags and invalid arguments. For example:
+as handling `--help` flags and invalid arguments.
+
+When using the [CommandRunner][] it replaces the [ArgParser][].
+
+In the following example we build a dart application called `dgit` that takes commands `commit` and `stash`.
+
+The [CommandRunner][] takes an 'executableName' which is ONLY used to generate the help message.
```
Comment (natebosch) [inline README.md:279]: I don't know why we need the emphasis on "only" here - it seem superfluous to me.

```suggestion
The [CommandRunner][] takes an `executableName` which is used to generate the help message.
```

File: README.md [hunk 3]
```diff
@@ -270,16 +270,39 @@ the left of `"commit"`, it is given to the top-level parser.
...
+
+> Note: for you final application to be called `dgit` rather than `dgit.dart` you would need to compile it using dart2native.
+
+e.g.
+`dgit commit -a`
+
+File `dgit.dart`
```
Comment (natebosch) [inline README.md:286]: Remove this line, I don't think we need to get into details like this - code sample is the focus here.

File: README.md [hunk 4]
```diff
@@ -270,16 +270,39 @@ the left of `"commit"`, it is given to the top-level parser.
...
 
 ```dart
-var runner = CommandRunner("git", "Distributed version control.")
-  ..addCommand(CommitCommand())
-  ..addCommand(StashCommand())
-  ..run(['commit', '-a']); // Calls [CommitCommand.run()]
+
+void main(List<String> args)
+{
```
Comment (natebosch) [inline README.md:291]:
```suggestion
void main(List<String> args) {
```

File: README.md [hunk 5]
```diff
@@ -270,16 +270,39 @@ the left of `"commit"`, it is given to the top-level parser.
...
+    ..addCommand(StashCommand())
+    ..run(args); 
 ```
 
-Custom commands are defined by extending the [Command][] class. For example:
+When the above `run(args)` line executes it parses the command line args looking for one of the commands (`commit` or `stash`).
+
+
+
+If the [CommandRunner][] finds a matching command then the [CommandRunner][] calls the overloaded `run()` method on the matching command (e.g. CommitCommand().run).
```
Comment (natebosch) [inline README.md:302]: "overloaded" isn't the right word here. "overridden" would be better, but I think unnecessary.

```suggestion
If the [CommandRunner][] finds a matching command then the [CommandRunner][] calls the `run()` method on the matching command (for example `CommitCommand().run`).
```

File: README.md (hunk 3 above)
Comment (bsutton) [reply README.md:286]: I don't really agree with this one.
Adding the command line usage improves comprehension.

During my first read through the doco I missed the code exampling showing (run(['commit', '-a')).

Showing the entire usage path is always useful and certainly does not harm.

File: README.md (hunk 1 above)
Comment (bsutton) [reply README.md:281]: removed.

**Review by natebosch — APPROVED** [review]

**Pull request merged**